### ISSUE TITLE
Enhancing editor completion for BaseModel.__init__

### ISSUE THREAD

**Issue by elda27** [samuelcolvin/pydantic#3475]
### Discussed in https://github.com/samuelcolvin/pydantic/discussions/3263

<div type='discussions-op-text'>

<sup>Originally posted by **elda27** September 26, 2021</sup>
Currently, `BaseModel.__init__` takes `**data` as an argument.
I would like to set `__annotations__`  for editor complementions.
I seem that `BaseModel.__init__.__annotations__` can be set in `ModelMetaclass`.

Is this change acceptable?
If it acceptable, I would like to contribute.</div>

I posted three months ago.
Arguments of `__init__` doesn't work code completion on the VSCode now.

**Current behavior**
![type_hint-before](https://user-images.githubusercontent.com/6873761/144431007-effb2062-b9e5-4000-b511-eb4dd11721ec.png)

**Expected proposing behavior**
![type_hint-after](https://user-images.githubusercontent.com/6873761/144431064-f16a28a0-814e-45d6-9a6a-6b4a6914ead8.png)

I think `__signature__` may work well since Python 3.9.
But practical annotation for `__init__` should be `__annotaions__` before Python 3.8.

I have already implemented this solution three months ago.
So I can contribute the solution.
Please consider my suggestion.

Following code is an example of changes at `ModelMetaclass` for code completion.
```python
if not hasattr(cls, '__init__'): # __init__ is defined at subclass
    cls.__init__.__annotations__ = {
        k: field.type_
        for k, field in fields.items()
    }
    cls.__init__.__kwdefaults__ = {
        k: field.default for k, field in fields.items()
    }
    cls.__init__.__kwdefaults__.update({
        field.alias: field.default
        for field in fields.values()
        if field.has_alias
    })
# set __signature__ attr only for model class, but not for its instances
cls.__signature__ = ClassAttribute('__signature__', 
generate_model_signature(cls.__init__, fields, config))
```

**Comment by shughes-uk** [comment]
This would be wonderful, dataclasses do this and it's incredibly helpful

**Comment by Bobronium** [comment]
Do you have a working PoC?

I don't think the approach you're going with will work. Correct me if I'm wrong, but IDE has no idea what `fields` in your code snippet are since it can only perform static analysis and won't run your code.

Also, this issue is addressed in https://github.com/samuelcolvin/pydantic/discussions/2698 and implemented in https://github.com/samuelcolvin/pydantic/pull/2721, so I would propose closing #3263 and this issye and continue discussion in https://github.com/samuelcolvin/pydantic/discussions/2698 for better visibility.

@shughes-uk, dataclasses don't do any of this. IDEs just have a custom behaviour defined to support them since they're included the the standard library.

**Comment by shughes-uk** [comment]
Dangit I had hoped that wasn't the case but had my suspicions 😭

**Comment by elda27** [comment]
@Bobronium 
I understand the reference issues. Thanks!

By the way,
I assume this code appending [`ModelMetaclass#L288`](https://github.com/samuelcolvin/pydantic/blob/415eb54f966f0eec7cb2fd0aa61492c17e606367/pydantic/main.py#L288).
Also, it was not a PoC, but something I had already implemented in my environment, and it was working. (If you insert the code I wrote below the line, you can see the completion as shown in the image.)
However, this is not working as expected yesterday. (Actually, the arguments shown seem to be the ones that I inserted into the __signature__.)

In particular, I'm using `pyright` (backend of `pylance`), which may be not static analyser.
I'm no still reading the implementation of `pyright`, so I may be wrong, but I think pyright analyzes metaclasses and decorators.
For example, `typing.TYPE_CHECKING` described in PEP564 is a variable created to target static analysis tools.
I thought the same was true for meta classes.

But I think there may be something wrong in what I am saying.
And I need to go read other PEPs to get a better understanding.

**Comment by Bobronium** [comment]
Wow, if it does really work, authors of `pyright` has done a really great job. It's is a static file checker as well.

However, I still can't wrap my head around how it could be implemented. Are you sure it picks up `__init__.__annotations__` and not class annotations that you define in body?

An easy check would be look like

```py
class AutocompleteMetaclass(ModelMetaclass):
    def __new__(*args, **kwargs):
        cls = super().__new__(*args, **kwargs)
        cls.__init__.__annotations__ = {'foo': int}`
        return cls

class Foo(FancyAutocompleteModel):
    bar: int

Foo(...)  # here you should have (bar: int) signature, if it really picks up annotations defined in metaclass
```

Also, keep in mind, that setting `__annotations__` like this doesn't make sense, since you're always changing annotations of `BaseModel.__init__`. You have to create new `__init__` for each model, which is not a trivial task. Here we struggled to do so and went with simpler solution of constructing and assigning a signature to the class: https://github.com/samuelcolvin/pydantic/pull/1034#discussion_r351925151

**Comment by miili** [comment]
Any update here?

**Comment by samuelcolvin** [comment]
#2721 is released with v1.9, I therefore think this is fixed.
